ipa-core/src/helpers/stream/chunks.rs

```diff
@@ -195,52 +193,208 @@ where
+    type Error = ();
+
+    fn try_from(value: &'a mut Option<T>) -> Result<Self, Self::Error> {
+        match value {
+            value @ Some(_) => Ok(DefinitelySome(value)),
+            None => Err(()),
+        }
+    }
+}
+
+impl<'a, T> DefinitelySome<'a, T> {
+    fn take(self) -> T {
+        self.0.take().unwrap()
+    }
+}
+
+impl<St, T, B, K, F, Fut, D, const N: usize> Stream
+    for StreamChunkProcessor<St, T, B, K, F, Fut, D, N>
+where
+    St: Stream<Item = Result<T, Error>> + Send,
+    B: ChunkBuffer<N, Item = T>,
+    K: IntoIterator,
+    F: Fn(usize, B::Chunk) -> Fut,
+    Fut: Future<Output = Result<K, Error>>,
+    D: Fn() -> T,
+{
+    type Item = MaybeChunkFuture<Fut, K, N>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.as_mut().project();
+
+        // dummy_fn is taken out of the `Option` when processing the final chunk, as an indicator
+        // that we should not poll the inner stream again.
+        let Ok(dummy_fn) = DefinitelySome::try_from(this.dummy_fn) else {
+            return Poll::Ready(None);
+        };
+
+        let (chunk_data, chunk_type) = loop {
+            match ready!(this.stream.as_mut().poll_next(cx)) {
+                Some(Ok(item)) => {
+                    this.buffer.push(item);
+                    if this.buffer.len() == N {
+                        break (this.buffer.take(), ChunkType::Full);
+                    }
+                }
+                Some(Err(e)) => {
+                    dummy_fn.take();
+                    return Poll::Ready(Some(MaybeFuture::value(Err(e))));
+                }
+                None if this.buffer.len() != 0 => {
+                    // Input stream ended, but we still have some items to process.
+                    let remainder_len = this.buffer.len();
+                    this.buffer.resize_with(N, dummy_fn.take());
+                    break (this.buffer.take(), ChunkType::Partial(remainder_len));
+                }
+                None => {
+                    // Input stream ended at a chunk boundary. Unlike the partial chunk case, we return
+                    // None, so we shouldn't be polled again regardless of dummy_fn signalling, but
+                    // might as well make this a fused stream since it's easy.
+                    dummy_fn.take();
+                    return Poll::Ready(None);
+                }
+            }
+        };
+        let idx = *this.pos;
+        *this.pos += 1;
+        Poll::Ready(Some(MaybeFuture::future(ChunkFuture::new(
+            (*this.process_fn)(idx, chunk_data.expect("ensured a full chunk")),
+            chunk_type,
+        ))))
+    }
+}
+
+/// Process stream through a function that operates on chunks.
+///
+/// Processes `stream` by collecting chunks of `N` items into `buffer`, then calling `process_fn`
+/// for each chunk. If there is a partial chunk at the end of the stream, `dummy_fn` is called
```

Comment:
Yes, I considered that -- it would require implementing `Default` on several layers, down to `AdditiveShare`. There is also a semantic required some places this is used that isn't exactly `Default` -- the dummy records need to contribute nothing to the overall result.

I think I would slightly prefer the version with `Default`, but I didn't want to presume to make that change given the impact.